tmt/steps/provision/mock.py

```diff
@@ -0,0 +1,505 @@
+import fcntl
+import os
+import io
+import select
+import shlex
+import subprocess
+from pathlib import Path
+from typing import Any, Optional, Union
+
+import mockbuild
+import mockbuild.config
+import mockbuild.plugins
+import mockbuild.plugins.pm_request
+
+import tmt
+import tmt.base
+import tmt.log
+import tmt.steps
+import tmt.steps.provision
+import tmt.utils
+from tmt.container import container, field
+from tmt.utils import Command, OnProcessStartCallback, Path, ShellScript
+from tmt.utils.wait import Waiting
+
+
+@container
+class MockGuestData(tmt.steps.provision.GuestData):
+    config: Optional[str] = field(
+        default=None,
+        option=('-r', '--config'),
+        metavar='CONFIG',
+        help='Mock chroot configuration file.',
+    )
+    rootdir: Optional[str] = field(
+        default=None,
+        option=('--rootdir'),
+        metavar='ROOTDIR',
+        help='The path for where the chroot should be built.',
+    )
+
+
+@container
+class ProvisionMockData(MockGuestData, tmt.steps.provision.ProvisionStepData):
+    pass
+
+
+class MockShell:
+    # Lazy object that collects chunks of bytes and decodes them when a
+    # newline is encountered
+    class Stream:
+        def __init__(self, logger):
+            self.logger = logger
+            self.output = bytes()
+            self.string = str()
+
+        def __iadd__(self, content: bytes):
+            self.output += content
+            while True:
+                pos = self.output.find(b'\n')
+                if pos == -1:
+                    break
+                string = self.output[:pos].decode('utf-8', errors='replace')
+                self.output = self.output[pos + 1:]
+                self.logger(string)
+                self.string += string
+                self.string += '\n'
+            return self
+
+    def __init__(self, parent: 'GuestMock', config: Optional[str], rootdir: Optional[str] = None):
```

Comment:
Both `config` and `rootdir` specified in the single instantiation of `MockShell`, is it necessary to have default value for one of them? `MockShell(parent=self, config=self.config, rootdir=self.rootdir)`.